### PR TITLE
G1.25 experiment content type updates

### PR DESCRIPTION
**Pull request by laceysanderson**
**Issue #25**

**Previous PR #26** (Closed due to the scope of changes and length of original)

## Motivation
We have determined that our original approach of having multiple Experiment Type content types (i.e. Field Experiment, Greenhouse Experiment, Biochemical Assay) does not work well since many experiments may include components of two experiment types. For example, the microbiome work involves sampling from a field experiment and then sequencing to determine the species composition. Another example of this would be a food quality experiment where we need to describe the biochem assay and the way the seeds were obtained.

## What does this PR do?

As such, this PR brings together all the fields from the original Experiment Type content types into a single content type. It also strives to organize these fields in a less overwhelming way and only have one copy of any common fields (i.e. name, objectives, summary).

More specifically, this PR

- [x] Removes the individual experiment type content types from `tripal.tripalentitytype_collection.trpcultivate_experiments.yml` in favour of a single `research_experiment` content type
- [x] Rearranges, names and combines any common fields across the 4 content types into a single content type in `tripal.tripalfield_collection.trpcultivate_experiments.yml` making sure not to loose any experiment type specific instructions.
- [x] Adds a default layout for the display (page view) of an experiment in `tripal_layout.tripal_layout_default_view.trpcultexperiment_layout.yml` to automate configuring the order of the many fields.
- [x] Adds a default layout for the form (edit) of an experiment in `tripal_layout.tripal_layout_default_form.trpcultexperiment_layout.yml` to automate configuring the order of the many fields.
- [x] Works to provide a consistent and intuitive organization of fields for both the experiment form and display.
- [x] Changes the "Germplasm Species" field to a "Germplasm Scientific Name" field to handle the situation where there are multiple genus'.
- [x] Add support for automatically creating markup and Entity Reference fields which cannot be added via the Tripal Fields Collection since they are not Tripal Fields.
- [x] Organize into vertical tabs to make the form less overwhelming
- [x] Change the organization to map better to how an experiment is designed with each tab representing steps of the process.
- [x] Add better group instructions for various panes.
- [x] Auto-configure some of the field widget size + options for a better experience (e.g. configure property to be a select list of options)
- [x] Auto-apply layouts.
- [x] Fix automated tests.

### What to Expect

Expand the following section to see screenshots of the form in order to see what to expect.

<details>
<summary>What to Expect</summary>

  Objectives
  Attribution
  Design
  Methodology
  References

</details>

### Manual Testing

1. Create a fresh container on this branch. If you are using VSCode devcontainers then just click "Reopen in container" and if not, see the instructions at the bottom.
2. Create a number of contacts by going to Top Bar > Tripal > Content > Add Tripal Content > Contact. We will use these as the researchers and organization in step 4.
3. Create a Grant and a study using the same approach.
4. Open the Create Research Experiment form by going to Top Bar > Tripal > Content > Add Tripal Content > Research Experiment
5. Check that the form layout is intuitive, fill it out and make sure it works this time 🙈 

#### Build container from scratch

```
cd ~/Dockers
git clone https://github.com/TripalCultivate/TripalCultivate trpcultBase-35
cd trpcultBase-35
git checkout g1.25-experimentContentTypeUpdates
docker build --tag=trpcultivate-base:review35 ./
docker run --publish=80:80 -tid --name=base35 --volume=`pwd`:/var/www/drupal/web/modules/contrib/TripalCultivate trpcultivate-base:review35
```

### CODE REVIEW

**Comment by ruobinLiu** [comment]
I am very happy with the updated version of the experiment content type as it allows more flexibility for those that include components of two experiment types!

In addition, it is very handy to have **Objectives**, **Attribution**, **Design**, **Methodology** and **Reference** separated and individually clickable on the side bar, so people don't get a super long list when they try to fill in details for an experiment. 

Plus, I really like how the "Save" button is moved to the top right instead of having to scroll all the way to the very bottom.

I forgot to fill in the URL alias at the beginning, then I was reminded when trying to save the page. The URL alias warning text is very helpful, so I now know I need to place a slash before the alias 😝  

Might also be helpful if Experiment Design section allows inserting of tables? 🤔 

Just using the microbiome work as an example:
Screenshot 2024-11-28 at 4 35 39 PM

Having the N movement greenhouse experiment as another example:
Screenshot 2024-11-28 at 4 38 21 PM
Screenshot 2024-11-28 at 4 56 56 PM


Nice work, thank you Lacey!

File: templates/markup_field/research_experiment.expdesign.twig [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+  <ul>
+    <li>biological and technical replication</li>
+    <li>randomization and blocking</li>
+    <li>treatments and controls</li>
+    <li>specific assay techniques used</li>
+    <li>sample or measurement collection considerations</li>
+  </ul>
+
+  <p>Make sure to not only state specific design/techniques but also to full describe why these choices were made and how they apply to your specific experiment. Also,remember that the above list is not meant to be exhaustive.</p>
```
Comment (carolyncaron) [inline templates/markup_field/research_experiment.expdesign.twig]:
```suggestion
  <p>Make sure to not only state specific design/techniques but also to fully describe why these choices were made and how they apply to your specific experiment. Also, remember that the above list is not meant to be exhaustive.</p>
```

File: templates/markup_field/research_experiment.timepoints.twig [hunk 2]
```diff
@@ -0,0 +1,10 @@
+<div class="instructions years">
+  <p>We want to include information about  the experiment was performed. <strong>For multiple dates use the "Add More" button to add a form element for each one and do NOT use ranges.</strong></p>
```
Comment (carolyncaron) [inline templates/markup_field/research_experiment.timepoints.twig]: This first sentence looked incomplete. Here's my suggestion to keep consistent with Locations:
```suggestion
  <p>Timepoints refers to the time period the experiment was performed, which depending on the experiment, may cover one or multiple growing seasons, or a single day. <strong>For multiple dates, use the "Add More" button to add a form element for each one and do NOT use ranges.</strong></p>
```

File: templates/markup_field/research_experiment.expdesign.twig [hunk 3]
```diff
@@ -0,0 +1,13 @@
+<div class="instructions experiment-design">
+  <p>Describe the design of your experiment in detail, making sure to include all the information needed to evaluate the design and for repeatability. We suggest including information on all of the following that apply to your type of experiment.</p>
```
Comment (carolyncaron) [inline templates/markup_field/research_experiment.expdesign.twig]:
```suggestion
  <p>Describe the design of your experiment in detail, making sure to include all the information needed to evaluate the design and for repeatability. We suggest including information on all of the following <strong>that apply to your type of experiment</strong>:</p>
```

**Review by carolyncaron — COMMENTED** [review]
The overall design looks great so far. 👏  The form filling experience is much more intuitive now! I'm very happy with the descriptions and particularly the fact that we can now bold important information using this approach 💯 

This is not a complete code review, but I did search for descriptions where I noticed typos or wanted to clarify, so I made suggestions there.

A few other things jumped out to me:
- Are there plans to have the "Genus" field autocomplete in the future? If not, I'm not sure why we need to have **both** Genus and Scientific name as options, unless we want to link to an organism in the database.
- I filled out every field in the form one tab at a time (ie. I saved between every tab and checked if the page contained what I expected), to try and simulate a real scenario of a Research Experiment being revisited before, during and after an experiment is completed. It went smoothly until I saved the `Design` tab, and noticed that all of my contacts in Attribution disappeared (all except for Research Grant):

Screenshot 2024-11-29 at 11 38 24 AM

   Things were smooth after that point though, and the Publication field did appear under Attribution after I chose "Unknown" and saved.

I'm going to hold back on more intensive testing until this PR is fully ready, unless you want the feedback sooner rather than later!

**Comment by laceysanderson** [comment]
Okay, I created issues for the following pointed out by Carolyn in her review as they are beyond the scope of this PR:

> Are there plans to have the "Genus" field autocomplete in the future? If not, I'm not sure why we need to have both Genus and Scientific name as options, unless we want to link to an organism in the database.

There is plans to create a field to make this a much simpler process and it is now documented in Issue #37. Essentially, we need both genus and scientific name because genus alone is needed for phenotypes configuration. The new field will allow the curator to only worry about selecting the organism (i.e. scientific name) and it will create the genus property in the background for phenotypes intengration.

> It went smoothly until I saved the Design tab, and noticed that all of my contacts in Attribution disappeared (all except for Research Grant)

The issue with the attribution is a known bug with the Contact By Role field that is prevasive. It needs to be fixed in Tripal Core but I haven't had a chance to do so yet or even to replicate in a pure Tripal core environment. As such there isn't an issue in Tripal core yet. That said, I'm now made an issue in this module to make sure it doesn't get lost: Issue #38.

**Review by carolyncaron — APPROVED** [review]
Thanks Lacey for creating those issues! I think this PR accomplishes what it set out to do. I plan to do a more thorough test of this form once there is more stability for fields in Tripal core. (Hint to self: the "Add more" functionality, being able to extensively edit the form, etc..)

**Pull request merged**